### PR TITLE
Make parent block selector visible and offset the toolbar.

### LINKED ISSUE #28522: Improve parent/child selection behavior in the Quote block
The Quote block is about to receive nesting support (#25892). This benefits transformations, and solves basic issues like being able to have images inside a quote.

The nesting change also makes it slightly harder to select the parent block. This ticket suggests that we build a system that allows the quote block to always show its "select parent" button, even when an inner-block is selected. GIF showing how this could work:

gif

This behavior could benefit blocks where the parent and immediate child are highly tied together, such as Quote, Navigation, Social Links, perhaps even Gallery.

### PR DESCRIPTION
**Pull request by mtias**
Closes #28522.
Complements #23800.

image

- [x] Show parent selector and offset toolbar.
- [x] Ensure toolbar keyboard navigation works properly.
- [x] Highlight parent block's outline on hover / focus.
- [x] Handle Top Toolbar.
- [x] Disable for mobile.

Parent highlighting outline:

image

Current block highlighting outline:

image

### CODE REVIEW

**Comment by jasmussen** [comment]
We've tried many things: the parent pop-out chip, clickthrough, breadcrumbs, the block traversal tool. Some are still around, others were not succesful. Because it's such an important tool, it's important to continue refining this interface.

And this feels like a natural evolution of the parent pop-out chip, probably why it's already been proposed and discussed. The benefits remain: 

- It's like a parent/child breadcrumb directly in the block toolbar.
- It's a good complement to existing tools.
 
Downsides remain: 

- It doesn't scale to mobile or the top toolbar option.
- It makes the block toolbar still wider.

However, since the original conversations, a few changes have been made which makes me think it's time to try it:

- The parent pop-out chip has not been obvious in practice.
- The block toolbar itself has had small enhancements to the grid and structure, with more still to come.
- Blocks like Social Links and Navigation have made strides in other aspects, but improvements to the parent/child relationship has become more clearly urgent.
- The Gallery and Quote blocks are due for block nesting soon, with a clear parent/child relationship being the primary challenge.

And in practice this change really does improve things. Notably the hover block boundary indication is highly helpful in indicating the behavior. Cover:

ex1

A paragraph inside a default group (fixes #20453):

paragraph

Columns benefit vastly:

columns

Navigation benefits:

navigation

I think there's still there's work to do separately. Notably the width of the block mover/drag control adds weight to child block toolbars, and in that the option doesn't work with the "top toolbar" option, it seems like there might be an opportunity to explore a solution that tackles both. Finally, blocks with already wide toolbars, such as List, will either need nesting (to split the long toolbar into parent and child) or other thought (group tools in dropdowns). 

But all of those followups are general benefits to the block UI, and ultimately the benefits of this approach seem worth trying. The "Group" block example above feels like a good litmus test: before it was unobvious that the paragraph was inside an invisible group, but with the added chip, its place in the hierarchy becomes immediately obvious, and will benefit both Quote and Gallery blocks as they get nesting.

🚀

File: packages/block-editor/src/components/block-toolbar/index.js [hunk 1]
```diff
@@ -79,6 +80,8 @@ export default function BlockToolbar( { hideDragHandle } ) {
 		}
 	);
 
+	// Account for the cases where the block toolbar is rendered within the
+	// header are and not contextually to the block.
```
Comment (mcsf) [inline packages/block-editor/src/components/block-toolbar/index.js:84]: Was this a typo for:

```suggestion
	// header area and not contextually to the block.
```

**Review by mcsf — APPROVED** [review]
Nice one.

File: packages/block-editor/src/components/block-list/block-contextual-toolbar.js [hunk 2]
```diff
@@ -12,29 +17,38 @@ import NavigableToolbar from '../navigable-toolbar';
...
 	}, [] );
 	if ( blockType ) {
 		if ( ! hasBlockSupport( blockType, '__experimentalToolbar', true ) ) {
 			return null;
 		}
 	}
+
+	// Shifts the toolbar to make room for the parent block selector.
+	const classes = classnames( 'block-editor-block-contextual-toolbar', {
+		'with-offset': hasParents,
```
Comment (talldan) [inline packages/block-editor/src/components/block-list/block-contextual-toolbar.js:44]:
```suggestion
		'has-offset': hasParents,
```

Minor, but the normal convention in the codebase is `has` or `is` to prefix a modifier. I think `has` probably makes the most sense

File: packages/block-editor/src/components/block-parent-selector/index.js [hunk 3]
```diff
@@ -54,13 +76,15 @@ export default function BlockParentSelector() {
...
+				ref={ nodeRef }
+				{ ...showMoversGestures }
 			>
 				<ToolbarButton
 					className="block-editor-block-parent-selector__button"
 					onClick={ () => selectBlock( firstParentClientId ) }
 					label={ sprintf(
 						/* translators: %s: Name of the block's parent. */
-						__( 'Select parent (%s)' ),
+						__( 'Select %s' ),
```
Comment (talldan) [inline packages/block-editor/src/components/block-parent-selector/index.js:87]: In #26135 this change was also proposed, but I don't think it's a good idea for situations where a block might be nested in same block type - e.g. a group within a group. That seems very confusing, I think it's probably quite confusing already with the icon being the same as the block icon.

Are there other options? I proposed just making it 'Select parent'. Alternatively could the text wrap in the text labels mode?

cc @tellthemachines.

File: packages/block-editor/src/components/block-toolbar/style.scss [hunk 4]
```diff
@@ -44,6 +44,16 @@
 	}
 }
 
+.block-editor-block-contextual-toolbar.with-offset {
```
Comment (talldan) [inline packages/block-editor/src/components/block-toolbar/style.scss:47]: To match the previous suggestion:
```suggestion
.block-editor-block-contextual-toolbar.has-offset {
```

File: packages/block-editor/src/components/block-parent-selector/index.js (hunk 3 above)
Comment (mtias) [reply packages/block-editor/src/components/block-parent-selector/index.js:87]: I think the same block nested within itself is enough of a special case that we might just as well treat that one separately. "Select parent" is rather technical and cumbersome for most cases (for example, selecting the Gallery that contains an Image, or the Quote that has a paragraph, etc). I think all of these cases benefit from a more straightforward phrasing.

File: packages/block-editor/src/components/block-list/block-contextual-toolbar.js (hunk 2 above)
Comment (mtias) [reply packages/block-editor/src/components/block-list/block-contextual-toolbar.js]: Yes, I keep trying to sneak in the `with-` convention :) Switching to `has-parent` as bit more clear.

**Comment by jasmussen** [comment]
🔥

**Comment by tellthemachines** [comment]
I'm guessing this change wasn't tested with text labels enabled:

Screen Shot 2021-02-04 at 12 22 57 pm

Text labels have variable widths, so fixing this will require removing the absolute positioning on the parent button, which also means we'll have to set the background and borders on the toolbar elements instead of the parent, to keep that little separation between parent button and the rest of the toolbar.

**Comment by jasmussen** [comment]
Sorry about that one, I'll try and follow with a PR.

**Comment by jasmussen** [comment]
Fix in https://github.com/WordPress/gutenberg/pull/28721.

**Comment by mtias** [comment]
Oh my, I was convinced we had not yet applied the text only treatment to the contextual block toolbar.

**Comment by tellthemachines** [comment]
I made sure to get it in for 5.7, as it felt weird to just have text labels in the top toolbar 😄

Thanks for the quick fix @jasmussen !

**Pull request merged**